core/sail/inferencer/src/main/java/org/eclipse/rdf4j/sail/inferencer/fc/FastRdfsForwardChainingSailConnetion.java

```diff
@@ -0,0 +1,1043 @@
+
+        calculateRangeDomain(fastRdfsForwardChainingSail.rangeStatements, fastRdfsForwardChainingSail.calculatedRange);
+        calculateRangeDomain(fastRdfsForwardChainingSail.domainStatements, fastRdfsForwardChainingSail.calculatedDomain);
+
+
+        fastRdfsForwardChainingSail.calculatedTypes.forEach((subClass, superClasses) -> {
+            addInferredStatement(subClass, RDFS.SUBCLASSOF, subClass);
+
+            superClasses.forEach(superClass -> {
+                addInferredStatement(subClass, RDFS.SUBCLASSOF, superClass);
+                addInferredStatement(superClass, RDFS.SUBCLASSOF, superClass);
+
+            });
+        });
+
+        fastRdfsForwardChainingSail.calculatedProperties.forEach((sub, sups) -> {
+            addInferredStatement(sub, RDFS.SUBPROPERTYOF, sub);
+
+            sups.forEach(sup -> {
+                addInferredStatement(sub, RDFS.SUBPROPERTYOF, sup);
+                addInferredStatement(sup, RDFS.SUBPROPERTYOF, sup);
+
+            });
+        });
+
+    }
+
+    private Set<Resource> resolveTypes(Resource value) {
+        Set<Resource> iris = fastRdfsForwardChainingSail.calculatedTypes.get(value);
+
+        return iris != null ? iris : Collections.emptySet();
+    }
+
+    private Set<IRI> resolveProperties(IRI predicate) {
+        Set<IRI> iris = fastRdfsForwardChainingSail.calculatedProperties.get(predicate);
+
+        return iris != null ? iris : Collections.emptySet();
+    }
+
+
+    private Set<Resource> resolveRangeTypes(IRI predicate) {
+        Set<Resource> iris = fastRdfsForwardChainingSail.calculatedRange.get(predicate);
+
+        return iris != null ? iris : Collections.emptySet();
+    }
+
+    private Set<Resource> resolveDomainTypes(IRI predicate) {
+        Set<Resource> iris = fastRdfsForwardChainingSail.calculatedDomain.get(predicate);
+
+        return iris != null ? iris : Collections.emptySet();
+    }
+
+
+    private void calculateSubClassOf(List<Statement> subClassOfStatements) {
+        subClassOfStatements.forEach(s -> {
+            Resource subClass = s.getSubject();
+            if (!fastRdfsForwardChainingSail.calculatedTypes.containsKey(subClass)) {
+                fastRdfsForwardChainingSail.calculatedTypes.put(subClass, new HashSet<>());
+            }
+
+            fastRdfsForwardChainingSail.calculatedTypes.get(subClass).add((Resource) s.getObject());
+
+        });
+
+        long prevSize = 0;
+        final long[] newSize = {-1};
+        while (prevSize != newSize[0]) {
+
+            prevSize = newSize[0];
+
+            newSize[0] = 0;
+
+            fastRdfsForwardChainingSail.calculatedTypes.forEach((key, value) -> {
+                List<Resource> temp = new ArrayList<Resource>();
+                value.forEach(superClass -> {
+                    temp
+                        .addAll(resolveTypes(superClass));
+                });
+
+                value.addAll(temp);
+                newSize[0] += value.size();
+            });
```

Comment:
Iteration until fixed point is not guaranteed to terminate in and of itself. The question comes down to if ever the result of resolveTypes can be infinite, or if it can decrease. 

It definitely wouldn't terminate if resolveTypes could remove perviously resolved types, e.g. negation. 

And it definitely wouldn't terminate if resolveTypes could generate new classes. 

So, it's guaranteed to terminate (barring any bugs) because there is always the fixed point where all things are subclasses of all things. 

Same goes for range and domain and sub-property.